dace/transformation/passes/analysis/analysis.py

```diff
@@ -28,25 +30,106 @@ def modifies(self) -> ppl.Modifies:
 
     def should_reapply(self, modified: ppl.Modifies) -> bool:
         # If anything was modified, reapply
-        return modified & ppl.Modifies.States
+        return modified & ppl.Modifies.CFG
+
+    def depends_on(self):
+        return {ControlFlowBlockReachability}
 
-    def apply_pass(self, top_sdfg: SDFG, _) -> Dict[int, Dict[SDFGState, Set[SDFGState]]]:
+    def apply_pass(self, top_sdfg: SDFG, pipeline_res: Dict) -> Dict[int, Dict[SDFGState, Set[SDFGState]]]:
         """
         :return: A dictionary mapping each state to its other reachable states.
         """
+        # Ensure control flow block reachability is run if not run within a pipeline.
+        if pipeline_res is None or not ControlFlowBlockReachability.__name__ in pipeline_res:
+            cf_block_reach_dict = ControlFlowBlockReachability().apply_pass(top_sdfg, {})
+        else:
+            cf_block_reach_dict = pipeline_res[ControlFlowBlockReachability.__name__]
         reachable: Dict[int, Dict[SDFGState, Set[SDFGState]]] = {}
         for sdfg in top_sdfg.all_sdfgs_recursive():
-            result: Dict[SDFGState, Set[SDFGState]] = {}
+            result: Dict[SDFGState, Set[SDFGState]] = defaultdict(set)
+            for state in sdfg.states():
+                for reached in cf_block_reach_dict[state.parent_graph.cfg_id][state]:
+                    if isinstance(reached, SDFGState):
+                        result[state].add(reached)
+            reachable[sdfg.cfg_id] = result
+        return reachable
+
+
+@properties.make_properties
+@transformation.experimental_cfg_block_compatible
+class ControlFlowBlockReachability(ppl.Pass):
+    """
+    Evaluates control flow block reachability (which control flow block can be executed after each control flow block)
+    """
+
+    CATEGORY: str = 'Analysis'
+
+    contain_to_single_level = properties.Property(dtype=bool, default=False)
+
+    def __init__(self, contain_to_single_level=False) -> None:
+        super().__init__()
+
+        self.contain_to_single_level = contain_to_single_level
 
+    def modifies(self) -> ppl.Modifies:
+        return ppl.Modifies.Nothing
+
+    def should_reapply(self, modified: ppl.Modifies) -> bool:
+        return modified & ppl.Modifies.CFG
+
+    def _region_closure(self, region: ControlFlowRegion,
+                        block_reach: Dict[int, Dict[ControlFlowBlock, Set[ControlFlowBlock]]]) -> Set[SDFGState]:
+        closure: Set[SDFGState] = set()
+        if isinstance(region, LoopRegion):
+            # Any point inside the loop may reach any other point inside the loop again.
+            # TODO(later): This is an overapproximation. A branch terminating in a break is excluded from this.
```

Comment:
fine for now. Thanks for leaving the comment

Reply:
Though it should be easy to add if breaks/returns are the only leaf reachable nodes? That's the only case where this is correct, otherwise the overapproximation is exact

Reply:
Yes that is correct - but there's not (yet) an easy API call to check if a nested region may break / exit the current loop context (coming in one of the following PRs), so checking whether the only reachable leaf breaks/returns is slightly more involved for me to suggest packing this into this PR. What do you think?

Reply:
next PR is good, I did not intend for it to be in this PR. We would need some kind of analysis test for it